### PR TITLE
[dtensor][BE] replace compute_local_shape with compute_local_shape_and_global_offset

### PR DESCRIPTION
**Pull request by XilunWu**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #135554

**Summary**
1. This PR removes the public API `compute_local_shape` and replace its use with the more general API `compute_local_shape_and_global_offset`. 
2. To keep `compute_local_shape_and_global_offset` consistent with `compute_local_shape` on empty shards, it now returns local tensor shape `(0,)` for empty shards which is more aligned with DTensor's semantics on non-participating ranks.

**Test**
`pytest test/distributed/_tensor/test_dtensor.py`
`pytest test/distributed/_tensor/test_init.py`
`pytest test/distributed/_tensor/test_tensor_ops.py`

cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @tianyu-l

Differential Revision: [D62415591](https://our.internmc.facebook.com/intern/diff/D62415591)

### CODE REVIEW

**Comment by XilunWu** [comment]
@XilunWu has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D62415591).

**Comment by XilunWu** [comment]
@XilunWu has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D62415591).

File: test/distributed/_tensor/test_utils.py [hunk 1]
```diff
@@ -27,36 +24,6 @@ class UtilTest(DTensorTestBase):
     def world_size(self):
         return 8
 
-    @with_comms
-    def test_compute_local_shape_2d_uneven(self):
```
Comment (tianyu-l) [inline test/distributed/_tensor/test_utils.py:31]: It seems here there are some interesting uneven cases. Should we keep them? E.g. we can just rename this function and switch function calls.

**Comment by wz337** [comment]
I think I did replace all the internal usage from compute_local_shape to compute_local_shape_and_global_offset, but it doesn't hurt to double check.

File: test/distributed/_tensor/test_utils.py (hunk 1 above)
Comment (XilunWu) [reply test/distributed/_tensor/test_utils.py:31]: merged those tests into `test_compute_local_shape_and_global_offset_2D`

**Comment by XilunWu** [comment]
@wz337 only 2-3 `compute_local_shape` call in DTensor internals which gets removed in this PR.

**Review by tianyu-l — APPROVED** [review]
SGTM

**Review by wz337 — APPROVED** [review]

**Pull request closed without merging**